### DISCUSSION TITLE
Can Popover be used as a controlled component?

### DISCUSSION THREAD

**Discussion by bstro** [chakra-ui/zag#500] [Q&A] [answered]
I'm looking at the docs and I don't see any `open` or `isOpen` prop in order to control the visibility of the popover from a parent. This is really useful for controlling popovers from things other than the trigger. Is this possible with the zag's popover?

my goal is to get the popover to open when the Editable input has focus, so my thought was to drive the popover as a controlled component.

**Comment by anubra266** [comment]
@bstro use open  [here](https://zagjs.com/overview/programmatic-control#transient-updates)

**Reply by bstro** [reply]
Thanks for the speedy reply! I looked at the link you provided about transient updates, and tried the following just to see if it worked (am aware that I'll need to memoize the actual value i end up using):
```
const [popoverState, popoverSend] = useMachine(
    popover.machine({
      id: createUniqueId(),
    }),
    { context: { open: true } }
  );
```
However, the types don't recognize any `open` properties on the context object. Is there something I'm missing?

**Reply by bstro** [reply]
I discovered `{ state: "open" }` for the second argument for `useMachine` but it doesn't seem to have any effect.

Here's a [stackblitz](https://stackblitz.com/edit/solidjs-templates-z2gd59?file=src%2Findex.tsx,src%2FInputWithPopover.tsx%3AL16). I'm trying to get the popover to open when the Editable input has focus.

**Comment by bstro** [comment]
I'm wondering if I was actually trying to solve the problem at the wrong level. I found that this approach allowed me to compose the machines together with the behavior I needed:
```
  const popoverTriggerId = createUniqueId();

  const [editableState, editableSend] = useMachine(
    editable.machine({
      id: createUniqueId(),
      ids: { input: popoverTriggerId },
      value: props.value,
      onChange: props.onChange,
      onEdit() {
        popoverApi().open();
      },
      onCancel() {
        popoverApi().close();
      },
    })
  );

  const editableApi = m(() =>
    editable.connect(editableState, editableSend, normalizeProps)
  );

  const [popoverState, popoverSend] = useMachine(
    popover.machine({
      id: createUniqueId(),
      ids: { trigger: popoverTriggerId },
    })
```
I don't understand why, but both the editable and popover machines must be initialized with `ids: { input: popoverTriggerId }` in the configuration object for this to work.

**Comment by anubra266** [comment, marked as answer]
@bstro You're doing it the right way now, I just realised we aren't using an `open` key in the context, so my first solution won't work.
But you shouldn't need the `ids` key to make it work
